srml/democracy/src/lib.rs

```diff
@@ -301,42 +301,34 @@ impl<T: Trait> Module<T> {
 					(Zero::zero(), bal * BalanceOf::<T>::sa(vote.multiplier() as u64), bal)
 				}
-			).fold((Zero::zero(), Zero::zero(), Zero::zero()), |(a, b, c), (d, e, f)| (a + d, b + e, c + f));
-		let (del_approve, del_against, del_capital) = Self::tally_delegation(ref_index);
-		(approve + del_approve, against + del_against, capital + del_capital)
+			).fold((Zero::zero(), Zero::zero(), Zero::zero()), |(a, b, c), (d, e, f)| (a + d, b + e, c + f))
 	}
 
 	/// Get the delegated voters for the current proposal.
 	/// I think this goes into a worker once https://github.com/paritytech/substrate/issues/1458 is done. 
-	fn tally_delegation(ref_index: ReferendumIndex) -> (BalanceOf<T>, BalanceOf<T>, BalanceOf<T>) {
-		Self::voters_for(ref_index).iter()
-			.fold((Zero::zero(), Zero::zero(), Zero::zero()), |(approve_acc, against_acc, capital_acc), voter| {
-				let vote = Self::vote_of((ref_index, voter.clone()));
-				let (votes, balance) = Self::delegated_votes(ref_index, voter.clone(), vote.multiplier(), MAX_RECURSION_LIMIT);
-				if vote.is_aye() {
-					(approve_acc + votes, against_acc, capital_acc + balance)
-				} else {
-					(approve_acc, against_acc + votes, capital_acc + balance)
-				}
-			})
+	fn add_delegated_votes(ref_index: ReferendumIndex) {
+		for voter in Self::voters_for(ref_index).iter() {
+			let vote = Self::vote_of((ref_index, voter.clone()));
+			Self::add_delegated_votes_to(ref_index, voter, &vote, MAX_RECURSION_LIMIT);
+		}
 	}
 
-	fn delegated_votes(
+	fn add_delegated_votes_to(
 		ref_index: ReferendumIndex,
-		to: T::AccountId,
-		min_lock_periods: LockPeriods,
+		to: &T::AccountId,
+		vote: &Vote,
 		recursion_limit: u32,
-	) -> (BalanceOf<T>, BalanceOf<T>) {
-		if recursion_limit == 0 { return (Zero::zero(), Zero::zero()); }
-		<Delegations<T>>::enumerate()
-			.filter(|(delegator, (delegate, _))| *delegate == to && !<VoteOf<T>>::exists(&(ref_index, delegator.clone())))
-			.fold((Zero::zero(), Zero::zero()), |(votes_acc, balance_acc), (delegator, (_delegate, periods))| {
-				let lock_periods = if min_lock_periods <= periods { min_lock_periods } else { periods };
-				let balance = T::Currency::total_balance(&delegator);
-				let votes = T::Currency::total_balance(&delegator) * BalanceOf::<T>::sa(lock_periods as u64);
-				let (del_votes, del_balance) = Self::delegated_votes(ref_index, delegator, lock_periods, recursion_limit - 1);
-				(votes_acc + votes + del_votes, balance_acc + balance + del_balance)
-			})
+	) {
+		if recursion_limit == 0 { return }
+		for (delegator, (delegate, periods)) in <Delegations<T>>::enumerate() {
+			if 	delegate == *to && !<VoteOf<T>>::exists(&(ref_index, delegator.clone())) {
+				let multiplier = if periods < vote.multiplier() { periods } else { vote.multiplier() };
+				let delegator_vote = Vote::new(vote.is_aye(), multiplier);
+				<VotersFor<T>>::mutate(ref_index, |voters| voters.push(delegator.clone()));
+				<VoteOf<T>>::insert(&(ref_index, delegator.clone()), delegator_vote);
```

Comment:
I'm not sure I understand the change, when and where from is `add_delegated_votes` supposed to be called?

1. In the offchain worker context we can't really alter the storage in a persistent fashion.
2. Afaict he check in the loop tests for `VoteOf::exists`, which may return `true` for a delegator, that got updated in the earlier iteration of this loop. Which pretty much breaks second-level delegation.
3. What if I change delegation after this function is called? It will leave the storage in a quite inconsistent state (`Vote` won't be cleared)
4. This logic should most likely be shared with `fn vote`

Reply:
The problem is that balances of delegators are not being locked (I messed up after a merge). 

I thought a clean way to do this, given the current implementation, would be to just transform the delegators into voters just before `tally` (using `add_delegated_votes` inside `bake_referendum`). So the rest of the code remains the same.

1 - I'm not sure how much the offchain worker can help given this.
2 - Second-level delegation should work because the loop tests for votes in the source  end of the edge (`delegator`) and the new voter is the target end (`delegate`).
3 - Yeah, the idea is to not do that, calling it immediately before `tally`.